### PR TITLE
Add uv sync hook to automatically sync envs after checkout/merge/rebase

### PR DESCRIPTION
**Pull request by UnknownPlatypus**
This is a reopen of #22 with comments from #24 taken into account.

This serves a similar purpose as the poetry-install [pre-commit hook](https://github.com/python-poetry/poetry/blob/main/.pre-commit-hooks.yaml#L17-L24) (and is inspired from it)
Closes https://github.com/astral-sh/uv-pre-commit/issues/17

### CODE REVIEW

**Review by dylwil3 — APPROVED** [review]
Wonderful, thank you!

**Comment by UnknownPlatypus** [comment]
@dylwil3 Btw, after experimenting a bit with this hook at work, we noticed a little quirk with the [`post-merge` git hook](https://git-scm.com/docs/githooks#_post_merge).
> This hook cannot affect the outcome of git merge and is not executed, if the merge failed due to conflicts.

So when you merge something with conflicts, you lose the automatic `uv sync` run sadly.
I don't know if you have ideas on how to solve this ? 


My first idea was to also configure a `uv sync` pre-commit hook similar to the current `uv lock` hook, that way it runs on the merge commit (and in general on any commit changing the pyproject/uv.lock files which seems reasonable?):

```yaml
- id: uv-sync-pre-commit
  name: uv-sync-pre-commit
  description: "Automatically run 'uv sync' on your project dependencies"
  entry: uv sync
  language: python
  files: ^(uv\.lock|pyproject\.toml|uv\.toml)$
  args: []
  pass_filenames: false
  minimum_pre_commit_version: "2.9.2"
``` 

I don't think it should blocked this MR but I'll be happy to open a followup to address this the way you think is the best

**Comment by lachaib** [comment]
> @dylwil3 Btw, after experimenting a bit with this hook at work, we noticed a little quirk with the [`post-merge` git hook](https://git-scm.com/docs/githooks#_post_merge).
> 
> > This hook cannot affect the outcome of git merge and is not executed, if the merge failed due to conflicts.
> 
> So when you merge something with conflicts, you lose the automatic `uv sync` run sadly. I don't know if you have ideas on how to solve this ?

Isn't executed then when running `git merge --continue`?

**Comment by dylwil3** [comment]
> So when you merge something with conflicts, you lose the automatic uv sync run sadly.

I think I'd rather leave the hook as-is, unless this becomes a common issue. I don't want it to conflict with the pre-commit hook. My feeling is that the main use-case for the hook in this PR is for pulling and switching. Pulls happen to involve either a merge or rebase, but should rarely have conflicts.

@zanieb I think this is in a good state to merge if you feel comfortable with it

**Comment by UnknownPlatypus** [comment]
@dylwil3 Agree, I'm fine with the current state of the hook and we will see if the merge conflict use case show up later on. 
I've just rebased my branch so it should be ready to merge

File: README.md [hunk 1]
```diff
@@ -89,6 +89,32 @@ To run the hook over multiple files at the same time:
...
+    # uv version.
+    rev: 0.5.9
+    hooks:
+      - id: uv-sync
+```
+In order to install this hook, you either need to specify `default_install_hook_types` (as above) and run `pre-commit install`, 
+or you have to install it via `pre-commit install --install-hooks -t post-checkout -t post-merge -t post-rewrite`.
+
+To synchronize all dependencies in a workspace:
+```yaml
```
Comment (zanieb) [inline README.md:110]:
```suggestion
To synchronize your dependencies upon branch checkout, pull or rebase:

```yaml
default_install_hook_types:
  - pre-commit
  - post-checkout
  - post-merge
  - post-rewrite
repos:
  - repo: https://github.com/astral-sh/uv-pre-commit
    # uv version.
    rev: 0.5.9
    hooks:
      - id: uv-sync
```

In order to install this hook, you either need to specify `default_install_hook_types` (as above) and run `pre-commit install`, 
or you have to install it via `pre-commit install --install-hooks -t post-checkout -t post-merge -t post-rewrite`.

To synchronize all dependencies in a workspace:

```yaml
```

Comment (zanieb) [inline README.md:110]:
```suggestion
To synchronize all dependencies in a workspace:

```yaml
```

File: README.md [hunk 2]
```diff
@@ -89,6 +89,32 @@ To run the hook over multiple files at the same time:
...
+  - post-merge
+  - post-rewrite
+repos:
+  - repo: https://github.com/astral-sh/uv-pre-commit
+    # uv version.
+    rev: 0.5.9
+    hooks:
+      - id: uv-sync
+```
+In order to install this hook, you either need to specify `default_install_hook_types` (as above) and run `pre-commit install`,
```
Comment (zanieb) [inline README.md:106]:
```suggestion
```

In order to install this hook, you either need to specify `default_install_hook_types` (as above) and run `pre-commit install`, 
```

File: README.md [hunk 3]
```diff
@@ -89,6 +89,32 @@ To run the hook over multiple files at the same time:
       files: ^requirements-dev\.(in|txt)$
 ```
 
+To synchronize your dependencies upon branch checkout, pull or rebase:
+```yaml
```
Comment (zanieb) [inline README.md:93]:
```suggestion
To synchronize your dependencies upon branch checkout, pull or rebase:

```yaml
```

File: README.md (hunk 2 above)
Comment (zanieb) [reply README.md:106]: _sigh_ GitHub.

I tried to fix the spacing in the suggestions but alas. Can you use newlines for consistency with the rest of the document?

File: .pre-commit-hooks.yaml [hunk 4]
```diff
@@ -28,3 +28,13 @@
...
+- id: uv-sync
+  name: uv-sync
+  description: "Automatically run 'uv sync' on your repository after a checkout, pull or rebase"
+  entry: uv sync
+  args: ["--locked"]
+  language: python
+  always_run: true
+  pass_filenames: false
+  stages: [post-checkout, post-merge, post-rewrite]
+  minimum_pre_commit_version: "2.9.2"
```
Comment (zanieb) [inline .pre-commit-hooks.yaml:40]: Should this have a trailing newline? I think so?

**Comment by UnknownPlatypus** [comment]
@zanieb I think I addressed all your comments, was not sure about the one that got weirdly formatted. Let me know if you need me to polish a few thing still. Thanks

**Comment by UnknownPlatypus** [comment]
I've just noticed with a coworker that this hook fail for people using keyring authentification via subprocess if keyring is not globally available.

Adding `additional_dependencies: [keyring]` ensure it always work. Should I add a mention in the readme ?

**Comment by zanieb** [comment]
> Adding additional_dependencies: [keyring] ensure it always work. Should I add a mention in the readme ?

That seems reasonable, but let's just do it in a separate pull request so we can get this merged.

**Review by zanieb — APPROVED** [review]
Thank you for your patience!

**Pull request merged**